src/gcp_airflow_foundations/source_class/ftp_gcs_source.py

```diff
@@ -0,0 +1,217 @@
+        TABLE_METADATA_FILE_{{ ds }}.csv
+    for a metadata file supplied daily.
+
+    The format of the metadata file should be a csv with one column as follows:
+        FILE_NAME_1
+        ...
+        FILE_NAME_N
+    with all files to ingest
+    """
+    source_type = "GCS"
+    gcs_hook = GCSHook()
+
+    def metadata_file_sensor(self, table_config, taskgroup):
+        """
+        Implements a sensor for the metadata file specified in the table config.
+        """
+        if "metadata_file" in table_config.extra_options.get("gcs_table_config"):
+            metadata_file_name = table_config.extra_options.get("gcs_table_config")["metadata_file"]
+            bucket = self.config.source.extra_options["gcs_bucket"]
+
+            return GCSObjectExistenceSensor(
+                task_id="wait_for_metadata_file",
+                bucket=bucket,
+                object=metadata_file_name,
+                task_group=taskgroup
+            )
+        else:
+            return DummyOperator(
+                task_id="dummy_metadata_file_ingestion",
+                task_group=taskgroup
+            )
+
+    def file_ingestion_task(self, table_config, taskgroup):
+        """
+        No ingestion is needed - data is already in GCS, so return a dummy operator.
+        """
+        return DummyOperator(
+            task_id="dummy_file_ingestion_operator",
+            task_group=taskgroup
+        )
+
+    def schema_file_sensor(self, table_config, taskgroup):
+        """
+        Implements an Airflow sensor to wait for an (optional) schema file in GCS
+        """
+        bucket = self.config.source.extra_options["gcs_bucket"]
+        schema_file_name = None
+        if "schema_file" in table_config.extra_options.get("gcs_table_config"):
+            schema_file_name = table_config.extra_options.get("gcs_table_config")["schema_file"]
+        logging.info(schema_file_name)
+
+        if schema_file_name:
+            return GCSObjectExistenceSensor(
+                task_id="wait_for_schema_file",
+                bucket=bucket,
+                object=schema_file_name,
+                task_group=taskgroup
+            )
+        else:
+            return DummyOperator(
+                task_id="dummy_schema_file_ingestion_operator",
+                task_group=taskgroup
+            )
+
+    def get_file_list_task(self, table_config, taskgroup):
+        return PythonOperator(
+            task_id="get_file_list",
+            op_kwargs={"table_config": table_config},
+            python_callable=self.get_list_of_files,
+            task_group=taskgroup
+        )
+
+    def file_sensor(self, table_config, taskgroup):
+        """
+        Returns an Airflow sensor that waits for the list of files specified by the metadata file provided.
+        """
+        bucket = self.config.source.extra_options["gcs_bucket"]
+        files_to_wait_for = "{{ ti.xcom_pull(key='file_list', task_ids='ftp_taskgroup.get_file_list') }}"
```

Comment:
why ftp?

Reply:
I woiuld make ftp_taskgroup a constant at the class level, so it is easier to change, instead of having it hardcoded